chasm/tree.go

```diff
@@ -926,3 +1088,64 @@ func newNode(
 		nodeName: nodeName,
 	}
 }
+
+func compareSideEffectTasks(a, b *persistencespb.ChasmComponentAttributes_Task) int {
+	if cmpResult := transitionhistory.Compare(a.VersionedTransition, b.VersionedTransition); cmpResult != 0 {
+		return cmpResult
+	}
+	return cmp.Compare(a.VersionedTransitionOffset, b.VersionedTransitionOffset)
+}
+
+func comparePureTasks(a, b *persistencespb.ChasmComponentAttributes_Task) int {
+	if cmpResult := a.ScheduledTime.AsTime().Compare(b.ScheduledTime.AsTime()); cmpResult != 0 {
+		return cmpResult
+	}
+
+	return compareSideEffectTasks(a, b)
+}
+
+func carryOverTaskStatus(
+	sourceTasks, targetTasks []*persistencespb.ChasmComponentAttributes_Task,
+	compareFn func(a, b *persistencespb.ChasmComponentAttributes_Task) int,
+) {
+	sourceIdx, targetIdx := 0, 0
+	for sourceIdx < len(sourceTasks) && targetIdx < len(targetTasks) {
+		sourceTask := sourceTasks[sourceIdx]
+		targetTask := targetTasks[targetIdx]
+
+		switch compareFn(sourceTask, targetTask) {
+		case 0:
+			targetTask.PhysicalTaskStatus = sourceTask.PhysicalTaskStatus
+			sourceIdx++
+			targetIdx++
+		case -1:
+			sourceIdx++
+		case 1:
+			targetIdx++
+		}
+	}
+
+	for ; targetIdx < len(targetTasks); targetIdx++ {
+		targetTasks[targetIdx].PhysicalTaskStatus = physicalTaskStatusNone
+	}
+}
+
+func taskCategory(
+	task *persistencespb.ChasmComponentAttributes_Task,
+) (tasks.Category, error) {
+	isImmeidate := task.ScheduledTime.AsTime().Equal(TaskScheduledTimeImmediate)
+
+	if task.Destination != "" {
+		if !isImmeidate {
+			return tasks.Category{}, serviceerror.NewInternal(
```

Comment:
Need outbound timer queue for this. 
Or maybe we can just run the task in timer queue if there's no strict destination isolation guarantee for the task.